Code repository related to manuscript "The impact of common variants on gene expression in the human brain: from RNA to protein to schizophrenia risk"

Following the run.sh in each content to run the code

# Requirements

## softwares and installation

listed in softwares name, version number, download link and installation instruction

     QTLTools v1.3.1 https://qtltools.github.io/qtltools/
     GNU parallel 20131022 https://ftp.gnu.org/gnu/parallel/ (if you use job scheduling system, like PBS, slurm, just ignore this software)
     MESC https://github.com/douglasyao/mesc/wiki
     PLINK v1.90p https://www.cog-genomics.org/plink2/
     GCTA v1.26.0 https://yanglab.westlake.edu.cn/software/gcta/#Download
     MR-JTI https://github.com/gamazonlab/MR-JTI/blob/master/model_training/predixcan/src/ (a part of codes used in 2.omics-specific-genes)
     MetaXcan v0.6.0 https://github.com/hakyimlab/MetaXcan (codes for 3.scz_risk_genes)
     python 3.6.3 https://www.python.org/downloads/release/python-363/
     R v4.0.3 https://cran.r-project.org/src/base/R-4/
     libraries: 
          qvalue 2.24.0; dplyr 1.0.10; ggplot2 3.4.0; ggrepel 0.9.1; ggplotify 0.1.0; cowplot 1.1.1; patchwork 1.1.1; ggpubr 0.4.0;
          ggVennDiagram 1.2.0; RColorBrewer 1.1.2; susieR 0.12.27; ivreg 0.6.1; TwoSampleMR 0.5.6; getopt 1.20.3; 
     
     
## datasets

listed in datasets name, download link

     gencode v19 annotation files https://www.gencodegenes.org/human/release_19.html
     GTEx BA9 eQTL https://storage.googleapis.com/adult-gtex/bulk-qtl/v7/single-tissue-cis-qtl/all_snp_gene_associations/Brain_Frontal_Cortex_BA9.allpairs.txt.gz 
     schizophrenia GWAS PGC3 https://figshare.com/articles/dataset/PGC3_SCZ_wave3_public_v2_GRCh37/19175744/1
     1000 Genome http://hgdownload.cse.ucsc.edu/gbdb/hg19/1000Genomes/phase3/

## files and results

     https://www.synapse.org/Synapse:syn50698163/files/

     
